### PR TITLE
Use plugin to mark node modules as external

### PR DESCRIPTION
**Pull request by leeweisberger**
We have been marking all node modules as external to prevent them from being bundled by reading the package.json. However, when using workspaces, there might be multiple package.jsons. Any packages referenced here are being bundled - this causes us to ship two versions of these deps - one in node_modules and one that is bundled into the source file.

Instead, use a plugin to mark all node_module deps as external. Maintaining behavior where we bundle in esm modules and ancestors of esm modules.

---

Question - why aren't we bundling node_modules? Wouldn't it just be easier (and produce a lighter container) if we just bundled node modules? If we did that, we could remove some of the complexity here and we could just ship a container with the shim.js file which contains everything needed to run the application.

### CODE REVIEW

File: pkg/build/node.go [hunk 1]
```diff
@@ -148,12 +135,6 @@ func node(root string, options KindOptions) (string, error) {
 		ARG BUILD_NPM_TOKEN
 		RUN [ -z "${BUILD_NPM_RC}" ] || echo "${BUILD_NPM_RC}" > .npmrc
 		RUN [ -z "${BUILD_NPM_TOKEN}" ] || echo "//registry.npmjs.org/:_authToken=${BUILD_NPM_TOKEN}" > .npmrc
-		# qemu (on m1 at least) segfaults while looking up a UID/GID for running
-		# postinstall scripts. We run as root with --unsafe-perm instead, skipping
-		# that lookup. Possibly could fix by building for linux/arm on m1 instead
-		# of always building for linux/amd64.
-		RUN npm install -g typescript@4.2 && \
```
Comment (leeweisberger) [inline pkg/build/node.go:155]: I don't think we need typescript? esbuild doesn't require it to build a typescript project.

File: pkg/build/node.go [hunk 2]
```diff
@@ -148,12 +135,6 @@ func node(root string, options KindOptions) (string, error) {
 		ARG BUILD_NPM_TOKEN
 		RUN [ -z "${BUILD_NPM_RC}" ] || echo "${BUILD_NPM_RC}" > .npmrc
 		RUN [ -z "${BUILD_NPM_TOKEN}" ] || echo "//registry.npmjs.org/:_authToken=${BUILD_NPM_TOKEN}" > .npmrc
-		# qemu (on m1 at least) segfaults while looking up a UID/GID for running
-		# postinstall scripts. We run as root with --unsafe-perm instead, skipping
-		# that lookup. Possibly could fix by building for linux/arm on m1 instead
-		# of always building for linux/amd64.
```
Comment (leeweisberger) [inline pkg/build/node.go:154]: How do we test for this? I am no longer using --unsafe-perm. Want to know if I am breaking something. Do we just have to test this on an m1 macbook?

File: pkg/build/node-build.mjs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+          const moduleRegexp = /node_modules\/(.+?)\//
+          const moduleMatch = resolved.match(moduleRegexp);
+          const module = moduleMatch[1];
+          if (esmModules.has(module)) {
+            esmPaths.add(resolved)
+            return resolved
+          }
+
+          return {
+            external: true,
```
Comment (colinking) [inline pkg/build/node-build.mjs:36]: Do we need to pass other `resolved` fields back? e.g. `{...resolved, external: true}`

File: pkg/build/node-build.mjs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+const args = process.argv.slice(2)
+const nodeVersion = args[0];
+build({
+  plugins: [
+    // This plugin marks all dependencies as external so they are not bundled 
+    // unless they are esm modules or packages pulled in by esm modules.
+    NodeResolvePlugin({
+      extensions: [".ts", ".js"],
+      onResolved: (resolved, importer) => {
+        if (esmPaths.has(importer)) {
```
Comment (colinking) [inline pkg/build/node-build.mjs:19]: Is `importer` exactly `node-fetch`, or should we be doing a regex match like below?

File: pkg/build/node-build.mjs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        return resolved;
+      },
+    }),
+  ],
+  entryPoints: ['/airplane/.airplane/shim.js'],
+  bundle: true,
+  platform: "node",
+  target: `node${nodeVersion}`,
+  outfile: "/airplane/.airplane/dist/shim.js",
+  external: ["airplane"],
```
Comment (colinking) [inline pkg/build/node-build.mjs:48]: I don't recall why airplane was external. Does it break anything if we bundle it?

File: pkg/build/node.go [hunk 6]
```diff
@@ -180,12 +161,8 @@ func node(root string, options KindOptions) (string, error) {
 		{{end}}
 		
 		RUN {{.InlineShim}} > /airplane/.airplane/shim.js && \
-			esbuild /airplane/.airplane/shim.js \
-				--bundle \
-				--external:airplane \
-				--platform=node {{.ExternalFlags}} \
-				--target=node{{.NodeVersion}} \
-				--outfile=/airplane/.airplane/dist/shim.js
+			{{.InlineBuild}} > /airplane/.airplane/build.mjs && \
```
Comment (colinking) [inline pkg/build/node.go:164]: Make sure to test this with `airplane dev`, too: https://github.com/airplanedev/lib/blob/612dba0540cfbcc76a51a6ec41de3c1cda3d77d7/pkg/runtime/javascript/javascript.go#L209

File: pkg/build/node.go (hunk 2 above)
Comment (colinking) [reply pkg/build/node.go:154]: I can give this a test for you.

**Comment by joshma** [comment]
(removing my review from here to unblock)

**Pull request closed without merging**